replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/WorkerManager.java

```diff
@@ -0,0 +1,123 @@
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Class responsible for managing item workers. It will monitor sites currently available in storage
+ * and create/destroy item workers accordingly for each site, respecting each site's parallelism
+ * factor.
+ */
+public class WorkerManager {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WorkerManager.class);
+
+  private final ScheduledExecutorService executorService = Executors
+      .newSingleThreadScheduledExecutor();
+
+  private final QueueBroker broker;
+
+  private final Set<String> siteIds;
+
+  private final Site localSite;
+
+  private final SiteManager siteManager;
+
+  private final NodeAdapters nodeAdapters;
+
+  private final Map<String, WorkerThreadPool> threadPools = new HashMap<>();
+
+  /**
+   * Creates a new worker manager.
+   *
+   * @param broker the queue broker for retrieving queues
+   * @param siteManager the site data access manager
+   * @param nodeAdapters node adapters factory
+   * @param properties replication properties
+   */
+  public WorkerManager(QueueBroker broker, SiteManager siteManager, NodeAdapters nodeAdapters,
+      ReplicationProperties properties) {
+    this.broker = broker;
+    this.siteManager = siteManager;
+    this.nodeAdapters = nodeAdapters;
+    this.siteIds = new HashSet<>(properties.getSites());
+    this.localSite = siteManager.get(properties.getLocalSite());
+  }
+
+  /**
+   * Initializes this manager and begins monitoring of sites.
+   */
+  public void init() {
+    executorService.scheduleAtFixedRate(this::monitorSites, 0L, 30L, TimeUnit.SECONDS);
+  }
+
+  private void monitorSites() {
+    if (siteIds.isEmpty()) {
+      try {
+        siteManager.objects().forEach(this::monitorSite);
+      } catch (ReplicationPersistenceException e) {
+        LOGGER.debug("Error retrieving sites", e);
+        LOGGER.error("Error retrieving sites to monitor, will try again next poll interval.");
+      }
+    } else {
+      for (String siteId : siteIds) {
+        try {
+          Site site = siteManager.get(siteId);
+          monitorSite(site);
+        } catch (NotFoundException e) {
+          WorkerThreadPool threadPool = threadPools.get(siteId);
+          if (threadPool != null) {
+            threadPool.shutdown();
+            threadPools.remove(siteId);
+          }
+          siteIds.remove(siteId);
+        } catch (ReplicationPersistenceException e) {
+          LOGGER.debug("Error retrieving site", e);
+          LOGGER.error("Error retrieving site {}. It will be retried next polling interval.", siteId);
+        }
+      }
+    }
+  }
+
+  private void monitorSite(Site site) {
+    final String siteId = site.getId();
+
+    final int poolSize = Math
+        .min(localSite.getParallelismFactor().orElse(1), site.getParallelismFactor().orElse(1));
+    WorkerThreadPool threadPool = threadPools.get(siteId);
```

Comment:
✏️ This can be achieve with:
```
threadPools.compute(siteId,this::createOrUpdate);

private WorkerThreadPool createOrUpdate(String site, @Nullable WorkerThreadPool current) {
   if (current == null) {
      return new WorkerThreadPool(broker.getQueue(siteId), poolSize, localSite, site,
          nodeAdapters);
   }
   current.setSize(poolSize);
   return current;
}
```